### PR TITLE
Feature/graph sketcher development

### PR DESCRIPTION
**Pull request by bh412**
* Adds '/sketcher' and implements Robin's marking of graph sketching questions 
* Limited to 3 lines at the moment
* Editable lines
* '/sketcher' acts as answer builder for content editors
* combines with relevant editor changes to enable graph question input

### CODE REVIEW

File: app/ts/inequality/Inequality.ts [hunk 1]
```diff
@@ -0,0 +1,683 @@
+/*
```
Comment (mlt47) [inline app/ts/inequality/Inequality.ts]: I'm guessing this file shouldn't be in this pull request?

File: app/js/app/router.js [hunk 2]
```diff
@@ -346,15 +346,15 @@ define(["angular", "@uirouter/angularjs"], function(angular, _angularUiRouter) {
         });
 
         // Temporarily disable until we have refactored
```
Comment (mlt47) [inline app/js/app/router.js]: We can remove the comment if it is no longer true

File: app/js/app/directives/graph_sketcher/GraphUtils.js [hunk 3]
```diff
@@ -1,22 +1,140 @@
...
+                else return 1;
+            }
+
+            clonedCurves.sort(compare);
+
+            function normalise(pt) {
+                let x = (pt[0] - canvasProperties.width/2) / canvasProperties.width;
+                let y = (canvasProperties.height/2 - pt[1]) / canvasProperties.height;
+                if (trunc) {
+                    pt[0] = Math.trunc(x * 10000) / 10000;
```
Comment (jsharkey13) [inline app/js/app/directives/graph_sketcher/GraphUtils.js]: And it is truncation, not rounding, we want to do here?

File: app/js/app/directives/graph_sketcher/GraphUtils.js [hunk 4]
```diff
@@ -89,28 +202,93 @@ define(function() {
...
+                    return prodrange / product(1, k);
+                }
+                prodrange = product(k + 1, n)
+                return prodrange / product(1, nMinusk);
+            }
+
+            let drawnNumberOfPoints = pts.length - 1;
+            let comb = [];
+            for (let currentIndex = 0; currentIndex <= drawnNumberOfPoints; currentIndex += 1) {
+                // from the other math library!!!! not the same as Math!!!!
```
Comment (jsharkey13) [inline app/js/app/directives/graph_sketcher/GraphUtils.js]:
```suggestion
```

We don't need this comment now that it's not `math.comb(...)`

File: app/js/app/directives/graph_sketcher/GraphUtils.js [hunk 5]
```diff
@@ -189,27 +348,28 @@ define(function() {
             return found;
         },
 
-        findEndPts: function(pts) {
+        findEndPts: function(pts) { //TODO look at this, pts now
```
Comment (jsharkey13) [inline app/js/app/directives/graph_sketcher/GraphUtils.js]: What's this TODO about?

File: app/js/app/directives/graph_sketcher/GraphUtils.js [hunk 6]
```diff
@@ -189,27 +348,28 @@ define(function() {
...
             if (pts.length == 0) return [];
 
             let ends = [];
 
-            ends.push(this.createPoint(pts[0].x, pts[0].y, pts[0].ind));
-            ends.push(this.createPoint(pts[pts.length - 2].x, pts[pts.length - 2].y, pts[pts.length - 2].ind));
+            ends.push(this.createPoint(pts[0][0], pts[0][1]));
+            ends.push(this.createPoint(pts[pts.length - 2][0], pts[pts.length - 2][1]));//TODO possible 1, from 101?
```
Comment (jsharkey13) [inline app/js/app/directives/graph_sketcher/GraphUtils.js]: Another unclear TODO?

Comment (jsharkey13) [reply app/js/app/directives/graph_sketcher/GraphUtils.js]: Why is it `length - 2` as well? Would be useful to justify that in a comment.

File: app/js/app/directives/graph_sketcher/GraphUtils.js [hunk 7]
```diff
@@ -189,27 +348,28 @@ define(function() {
...
+
 
             for (let i = 1; i < pts.length; i++) {
-                if (pts[i-1].x - pts[i].x > 200) {
-                    ends.push(this.createPoint(pts[i-1].x, pts[i-1].y, pts[i-1].ind));
-                    ends.push(this.createPoint(pts[i].x, pts[i].y, pts[i].ind));
+                if (pts[i-1][0] - pts[i][0] > 200) {
```
Comment (jsharkey13) [inline app/js/app/directives/graph_sketcher/GraphUtils.js]: This 200 is also a magic number?

File: app/js/app/directives/graph_sketcher/GraphUtils.js [hunk 8]
```diff
@@ -271,60 +431,71 @@ define(function() {
...
 
+            let position = null
 
+            for (let i = 0; i < statPts.length; i++) { 
+                for (let j = 0; j < pts.length; j++) {
+                    if (statPts[i][0] == pts[j][0]) {
+                        position = j;
                     }
                 }
+                ((statPts[i][1] < pts[position-5][1] && statPts[i][1] < pts[position+5][1])) && pot_max.push(statPts[i]);
```
Comment (jsharkey13) [inline app/js/app/directives/graph_sketcher/GraphUtils.js]: This is really an `if` statement, isnt it? There's no need to obscure it to save space; the minifier will do that fine!

File: app/js/app/directives/graph_sketcher/GraphUtils.js [hunk 9]
```diff
@@ -271,60 +431,71 @@ define(function() {
...
 
+        duplicateStationaryPts: function(pts, mode) {
+            let non_duplicates = []
+            for (let i = 0; i < pts.length; i++) {
+                let similar_ind = [pts[i]]
+                for (let j = 0; j < pts.length; j++) {
+                    (pts[j][0] !== pts[i][0]) && ((pts[j][0] < pts[i][0] + 5) && (pts[j][0] > pts[i][0] - 5)) && similar_ind.push(pts[j])
+                }
+                mode == 'maxima' ? similar_ind.sort(function(a, b){return a[1] - b[1]}) : similar_ind.sort(function(a, b){return b[1] - a[1]})
+                non_duplicates.indexOf(similar_ind[0]) === -1 ? non_duplicates.push(similar_ind[0]) : {};
```
Comment (jsharkey13) [inline app/js/app/directives/graph_sketcher/GraphUtils.js]: These are also `if` statements, and shouldn't be ternary statements

File: app/js/app/directives/graph_sketcher/GraphUtils.js [hunk 10]
```diff
@@ -487,30 +644,31 @@ define(function() {
...
                 selectedCurve.minima = this.findTurnPts(selectedCurve.pts, 'minima');
-                let minX = selectedCurve.pts[0].x;
-                let maxX = selectedCurve.pts[0].x;
-                let minY = selectedCurve.pts[0].y;
-                let maxY = selectedCurve.pts[0].y;
+                let minX = selectedCurve.pts[0][0];
+                let maxX = selectedCurve.pts[0][0];
+                let minY = selectedCurve.pts[0][1];
+                let maxY = selectedCurve.pts[0][1];
                 for (let k = 1; k < selectedCurve.pts.length; k++) { // TODO BH search through 'important' points instead
```
Comment (jsharkey13) [inline app/js/app/directives/graph_sketcher/GraphUtils.js]: Should this TODO be actioned or removed?

File: app/js/app/directives/graph_sketcher/GraphView.js [hunk 11]
```diff
@@ -34,14 +34,14 @@ define(['./GraphUtils.js'], function(graphUtils) {
             // want to connect closest points x,y wise, not just x wise
             let pts = curve.pts;
             for (let i = 1; i < pts.length; i++) {
-                if (pts[i].x - pts[i-1].x < 100 && pts[i].y - pts[i-1].y < 100) {
-                    this.p.line(pts[i-1].x, pts[i-1].y, pts[i].x, pts[i].y);
+                if (pts[i][0] - pts[i-1][0] < 100 && pts[i][1] - pts[i-1][1] < 100) {// TODO why <100?
```
Comment (jsharkey13) [inline app/js/app/directives/graph_sketcher/GraphView.js]: Another magic number, at least with a TODO this time; replace TODO with explanation.

File: app/js/app/services/QuestionActions.js [hunk 12]
```diff
@@ -51,10 +51,10 @@ define([], function() {
...
 					}
 
-					if (scope.doc.type == "isaacGraphSketcherQuestion") {
-						scope.question.validationResponse = TEMP_GRAPH_SKETCHER_RESULT;
-						return; // Do not submit Graph SKetecher question attemps
-					}
+					// if (scope.doc.type == "isaacGraphSketcherQuestion") {
+					// 	scope.question.validationResponse = TEMP_GRAPH_SKETCHER_RESULT;
+					// 	return; // Do not submit Graph SKetecher question attemps
+					// }
```
Comment (jsharkey13) [inline app/js/app/services/QuestionActions.js]: This should be removed if the questions are going to be checked properly now.

File: app/js/app/directives/graph_sketcher/GraphPreview.js [hunk 13]
```diff
@@ -14,15 +14,29 @@ define(["p5", "./GraphView.js", "./GraphUtils.js", "/partials/graph_sketcher/gra
             link: function(scope, element, _attrs) {
                 let graphPreviewDiv = element.find(".graph-preview");
 
-                scope.canvasID = scope.questionDoc.id;
+                // if (scope.state.curves == undefined || scope.state.curves == []) {
+                //     scope.canvasID = undefined;
+                // }
+
+                if(typeof scope.canvasID !== "undefined") {
+                    scope.canvasID = scope.questionDoc.id;
```
Comment (Morpheu5) [inline app/js/app/directives/graph_sketcher/GraphPreview.js]: Is `questionDoc.id` a number?

File: app/js/app/directives/graph_sketcher/GraphUtils.js [hunk 14]
```diff
@@ -1,22 +1,138 @@
...
+                clonedCurves[i].minX = Math.trunc(tmp * 1000) / 1000;
+
+                tmp = (clonedCurves[i].maxX - canvasProperties.width/2) / canvasProperties.width;
+                clonedCurves[i].maxX = Math.trunc(tmp * 1000) / 1000;
+
+                tmp = (canvasProperties.height/2 - clonedCurves[i].minY) / canvasProperties.height;
+                clonedCurves[i].minY = Math.trunc(tmp * 1000) / 1000;
+
+                tmp = (canvasProperties.height/2 - clonedCurves[i].maxY) / canvasProperties.height;
+                clonedCurves[i].maxY = Math.trunc(tmp * 1000) / 1000;
```
Comment (Morpheu5) [inline app/js/app/directives/graph_sketcher/GraphUtils.js]: Magic numbers. See also L60-61.

**Pull request merged**